samples/hssi/hssi.cpp

```diff
@@ -0,0 +1,654 @@
+#define CSR_RND_SEED1       0x3c0b
+#define CSR_RND_SEED2       0x3c0c
+#define CSR_PACKET_LENGTH   0x3c0d
+
+#define CSR_NUM_PKT         0x3d00
+#define CSR_PKT_GOOD        0x3d01
+#define CSR_PKT_BAD         0x3d02
+#define CSR_AVST_RX_ERR     0x3d07
+
+#define CSR_MAC_LOOP        0x3e00
+
+#define MBOX_TIMEOUT 1000
+
+bool parse_filter_address(properties::ptr_t filter, const char *sbdf)
+{
+  regex_t re;
+  regmatch_t matches[5];
+
+  memset(&matches, 0, sizeof(matches));
+
+  std::string sbdf_expr("([0-9a-fA-F]{4}):([0-9a-fA-F]{2}):([0-9a-fA-F]{2})\\.([0-7])");
+
+  regcomp(&re, sbdf_expr.c_str(), REG_EXTENDED|REG_ICASE);
+  bool match = regexec(&re, sbdf, 5, matches, 0) == 0;
+
+  if (match) {
+    std::string segment(sbdf + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so);
+    filter->segment = static_cast<uint16_t>(std::stoul(segment, nullptr, 16));
+
+    std::string bus(sbdf + matches[2].rm_so, matches[2].rm_eo - matches[2].rm_so);
+    filter->bus = static_cast<uint8_t>(std::stoul(bus, nullptr, 16));
+
+    std::string device(sbdf + matches[3].rm_so, matches[3].rm_eo - matches[3].rm_so);
+    filter->device = static_cast<uint8_t>(std::stoul(device, nullptr, 16));
+
+    std::string fn(sbdf + matches[4].rm_so, matches[4].rm_eo - matches[4].rm_so);
+    filter->function = static_cast<uint8_t>(std::stoul(fn, nullptr, 16));
+
+    regfree(&re);
+    return true;
+  }
+
+  regfree(&re);
+  memset(&matches, 0, sizeof(matches));
+
+  std::string bdf_expr("([0-9a-fA-F]{2}):([0-9a-fA-F]{2})\\.([0-7])");
+
+  regcomp(&re, bdf_expr.c_str(), REG_EXTENDED|REG_ICASE);
+  match = regexec(&re, sbdf, 5, matches, 0) == 0;
+
+  if (match) {
+    filter->segment = 0;
+
+    std::string bus(sbdf + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so);
+    filter->bus = static_cast<uint8_t>(std::stoul(bus, nullptr, 16));
+
+    std::string device(sbdf + matches[2].rm_so, matches[2].rm_eo - matches[2].rm_so);
+    filter->device = static_cast<uint8_t>(std::stoul(device, nullptr, 16));
+
+    std::string fn(sbdf + matches[3].rm_so, matches[3].rm_eo - matches[3].rm_so);
+    filter->function = static_cast<uint8_t>(std::stoul(fn, nullptr, 16));
+  }
+
+  regfree(&re);
+  return match;
+}
+
+handle::ptr_t open_accelerator(const char *guid, const char *sbdf)
+{
+  handle::ptr_t accel;
+
+  auto filter = properties::get();
+  filter->type = FPGA_ACCELERATOR;
+  filter->guid.parse(guid);
+
+  if (sbdf)
+    if (!parse_filter_address(filter, sbdf)) {
+      std::cerr << "Invalid PCIe address: " << sbdf << std::endl;
+      return accel;
+    }
+
+  auto tokens = token::enumerate({filter});
+  if (tokens.size() < 1) {
+    std::cerr << "Accelerator not found." << std::endl;
+    return accel;
+  }
+
+  if (tokens.size() > 1) {
+    std::cerr << "warning: More than one accelerator found." << std::endl;
+  }
+
+  return handle::open(tokens[0], 0);
+}
+
+#define INVALID_MAC 0xffffffffffffffffULL
+uint64_t mac_string_to_bits(const std::string &str)
```

Comment:
Are there any Linux calls to do this?

Reply:
I looked but didn't find any.  Maybe @fpgamatt can point one out?  There are routines for the IPv4 dotted quad conversion, but didn't see any for ethernet addresses.

Reply:
ether_ntoa() and ether_aton()

Reply:
Thanks